### PR TITLE
feat(dns): Rewrite dns persistence to allow virtual-outbound to be added

### PR DESCRIPTION
**Pull request by lahabana**
Currently the persisted state on the vips was very restraining.
With the addition of virtual-outbounds we'll need something more flexible
This migrates the state to a new format that will be extensible.

It also avoids reusing vips straight away which might avoid some issues when services are added and removed at once.

### CODE REVIEW

File: pkg/dns/vips/virtual_outbound.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	for entry := range vo.byHostname {
+		if _, ok := new.byHostname[entry]; !ok {
+			changes = append(changes, Change{Type: Remove, Entry: entry})
+		}
+	}
+	for entry, vob := range new.byHostname {
+		oldVob, ok := vo.byHostname[entry]
+		if ok {
+			if !oldVob.Equal(vob) {
+				changes = append(changes, Change{Type: Modify, Entry: entry})
```
Comment (lobkovilya) [inline pkg/dns/vips/virtual_outbound.go:112]: How `Modify` can happen?

File: pkg/dns/vips/virtual_outbound.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+func (mo *MeshOutbound) String() string {
+	var tags []string
+	for k, v := range mo.TagSet {
+		tags = append(tags, fmt.Sprintf("%s=%s", k, v))
+	}
+	sort.Strings(tags)
+	return fmt.Sprintf("%s=%d{%s}", mo.Origin, mo.Port, strings.Join(tags, ","))
+}
+
+type VirtualOutboundView struct {
```
Comment (lobkovilya) [inline pkg/dns/vips/virtual_outbound.go:48]: If this is opposed to GlobalView, I'd use `mesh` somewhere in the name. What do you think?

**Review by lobkovilya — COMMENTED** [review]
That looks great! I guess it's a lot of hard work to refactor/improve this part of the codebase. Thanks! I have few comments, mostly about naming, but overall it totally makes sense to me

File: pkg/dns/vips/virtual_outbound.go (hunk 1 above)
Comment (lahabana) [reply pkg/dns/vips/virtual_outbound.go:112]: When we add a new outbound (for example same host but different port).

**Review by lobkovilya — APPROVED** [review]

**Pull request merged**